### ISSUE TITLE
Reproduce favorita benchmark with tft

### ISSUE THREAD

**Issue by asmaier** [google-research/google-research#1799]
For everyone struggling with running and reproducing results of the TFT paper I leave this information here on how to reproduce the favorita benchmark with TFT:

1. Get Kaggle account
    - Phone verify Kaggle account
    - Join favorita competition and understand and accept rules
    - Create an API token and download the kaggle.json file
    - Create `.kaggle` directory and move `kaggle.json` file inside with the correct rights

            $ mkdir ~/.kaggle 
            $ mv kaggle.json ~/.kaggle
            $ chmod 600 ~/.kaggle/kaggle.json      
          
2. On a Ubuntu 20 x86 machine (Mac M1 won't work, because tensorflow 1.x will not be supported)
    1. Install pyenv
        - see https://brain2life.hashnodeq.dev/how-to-install-pyenv-python-version-manager-on-ubuntu-2004 
        - add the following to `~/.profile` (`.bashrc` doesn't work on Ubuntu !)

                export PYENV_ROOT="$HOME/.pyenv"
                command -v pyenv >/dev/null || export PATH="$PYENV_ROOT/bin:$PATH"
                eval "$(pyenv init -)"
     
    1. Install python 3.7 with pyenv

            $ pyenv install 3.7
            $ pyenv global 3.7.17
            
    1. Download tft directory
    
            $ svn export https://github.com/google-research/google-research/trunk/tft
            
    4. Create output repo
    
            $ mkdir -p ~/tft_outputs  
            
    5. in `tft` directory edit `requirement.txt` , add `protobuf<3.21` and `kaggle` and pin `pandas==0.25.3`
    
            numpy>=1.17.4
            pandas==0.25.3
            scikit-learn>=0.22
            tensorflow-probability>=0.8.0
            wget>=3.2
            pyunpack>=0.1.2
            patool>=1.12
            protobuf<3.21
            kaggle

    6. Edit `run.sh` and set `favorita` experiment

            # Expt options include {volatility, electricity, traffic, favorita}
            EXPT=favorita

    7. Run the `run.sh` script using `bash` (Don't use `sh`, it won't work)

           $ bash run.sh

   8. The script will fail, because it cannot find the favorita dataset. We have to download it manually. But the script created a venv and the `data` dir for us, which we will use now

            $ source ~/tft_outputs/venv/bin/activate   # use the venv generated by run script with the kaggle dependency installed
            $ cd ~/tft_outputs/data
            $ kaggle competitions download -c favorita-grocery-sales-forecasting

   9. Run the `run.sh`script again. Hopefully it will be the final run

        see https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike
   
            Passing list-likes to .loc or [] with any missing label will raise
            KeyError in the future, you can use .reindex() as an alternative.

    10. Sorry, it won't be the last time. This was just a test run. To really start the full training set `use_testing_mode=False` in the last line of `script_train_fixed_params.py`

            main(
                expt_name=name,
                use_gpu=use_tensorflow_with_gpu,
                model_folder=os.path.join(config.model_folder, "fixed"),
                data_csv_path=config.data_csv_path,
                data_formatter=formatter,
                use_testing_mode=False)  # Change to false to use original default params

    11. Start the real training run with

            $ nohup python3 -u script_train_fixed_params.py favorita ~/tft_outputs/ no 2>&1 >tft_output.txt &

        The `-u` option will make sure `python` is generating output unbuffered, so you won't get confused, why not all output shows up in the output file `tft_output.txt` immediatelly. 
        
**NOTE**: Be aware that the training isn't using all the data but only a sub sample of 450000 data point, as hardcoded in https://github.com/google-research/google-research/blob/ae1df11631983953cf335f7b91663577c76ae3b5/tft/data_formatters/favorita.py#L284

To use all the training data you have to change this method. However training TFT on the full favorita data set takes a really long time even with GPU support and will not improve the results by much. 

Another point to note is that the good benchmark results of TFT are dependent on the fact that in the preprocessing code NAN values of the log_sales feature are replaced by the mean value, see https://github.com/google-research/google-research/blob/ae1df11631983953cf335f7b91663577c76ae3b5/tft/data_formatters/favorita.py#L218
Choosing a different imputation method (e.g. replacing with zeros) makes the benchmark results of TFT much worse.

**Comment by DonLeif** [comment]
Thanks for this extremely helpful writeup.
I am currently going through the code myself and have some questions, if you allow.
1. I am seeing no seed anywhere except for when producing the hyperparam list. How can results be reproducible?
2. Training uses 450k samples. Sampling 450k datapoints from the 2015-04-01 (encoder length of 90 days) to 2015-12-01 period gives us roughly 1.6% of data and covers roughly 86% of all available time series (as per traj_id). When preprocessing data, the authors exclude cold-start items that are launched only after the training period. They do not, however remove the approx 14% of time series the model has never seen during training. Do I understand correctly? Isn't that a bit inconsistent?
Why wouldn't you decide on a subset of time series that is relevant and train/test only with these?

**Comment by asmaier** [comment]
Very valid questions. Let me try to remember, because its been a while since I worked with TFT. 

1. I believe setting `np.random.seed` somewhere in the code will actually fix the global state of numpy random generator, see https://stackoverflow.com/a/5836372/179014. So that might be the reason why the results are in fact reproducible. At least I was able to reproduce the results from the publication to several digits following the documentation above.

2. You might be right and this is inconsistent. I haven't looked that deeply into their results. I was happy when I was able to reproduce the results from the paper. But if I understand correctly you are saying that they exclude certain data from training, but don't exclude it from testing? I think by doing so they might have accidentally made their results a bit worse. You could try to remove the 14% from the test set and see how it changes the result. Maybe the results are a bit better. But I don't expect to see a huge improvement, because even training on the full data set didn't improve the results by much. At some point even the smartest algorithm has extracted all the information that is available in the given data and no further improvement is possible. All you can do then is collect more data.
